lib/route/ROUTING.md

```diff
@@ -0,0 +1,446 @@
+```
+
+Matching on the first segment is done by using the literal value "user" as the first argument to
+the constructor that we added to the `case` expression:
+
+```haskell
+  FrontendRoute_User -> PathSegment "user" _todo
+```
+
+We can now move onto building the `Encoder` for our `UserId`.
+
+#### Wait a minute, shouldn't I be matching on the route on the left?
+
+In many routing systems you match directly on the route input and break it down as required. This
+process can be fragile and has limitations. Additionally you are left on your own when comes to
+_creating_ links in your application. Because there is no way to relate the structure of a route
+to anything. You have to manually build up routes again, and if those routes change it can be
+onerous to find and fix all the constructed links.
+
+Obelisk routes are bidirectional, which means the `Encoder` that you create works as both a
+pattern match for incoming routes. As well as using the GADT constructors as a type safe
+mechanism for _creating_ links in your application. It's a compile error to try to use route
+constructors that don't exist, and if you change the type of a route the application will not
+build until you fix that change every where it appears.
+
+This design also enables the composition of `Encoder`s which is difficult and much more prone to
+error if you start by matching on the route itself.
+
+----
+
+### UserId Encoder
+
+Because the Obelisk routes are bidirectional, we need describe how to encode our type as a url
+segment at the same time as describing how to encode the url segment as our type.
+
+In other words, we need to tell Obelisk how to create a `UserId` from `Text` and back again. To
+do this we're going to write two functions and put them together into a `Prism'`.
+
+A `Prism'` is a condensed way of describing how to go from one type to another and _maybe_ back
+the other way.
+
+The types of the two functions are:
+
+```haskell
+_ :: UserId -> Text
+-- and
+_ :: Text -> Either Text UserId -- because not all Text is a valid UserId
+```
+
+The two functions are combined into a `Prism'` using the `prism` function. That function and the
+`Prism'` type will need to be imported from the `Control.Lens` module.
+
+```haskell
+import Control.Lens (Prism', prism)
+```
+
+Turning our `UserId` into `Text` is unremakrable so let's do that one first. Define the prism in
+the `Common.Api` module, next to our `UserId` newtype definition:
+
+```haskell
+userIdTextPrism :: Prism' Text UserId
+userIdTextPrism = prism toText _fromText
+  where
+    -- Unwrap the UserId, use 'show' on the 'Int' value, and then turn that into 'Text'
+    toText = T.pack . show . unUserId
+```
+
+Next to encode `Text` from the route input as a `UserId`. This means we have to check that this
+input is valid. Our requirement is that the input is a number greater than zero:
+
+```haskell
+userIdTextPrism :: Prism' Text UserId
+userIdTextPrism = prism toText fromText
```

Comment:
This isn't a good approach to implementing this function.  What we should do (and teach users to do) is to break down their encoders into small pieces, and then compose those.

What this is doing right now is 3 steps: unwrapping the UserId, showing the contents, and packing the resulting string as text.  The *combination* of those three things had better round-trip with the *combination* of the three parsing steps that are listed below.  This is dangerous - and, in fact, as a result, you've written a bug!  `UserId (-1)` does not roundtrip, which means that this prism, and the encoder based on it, are both invalid.

Instead, the correct way to implement it is to write each tiny piece of roundtripping logic separately, and then compose the resulting `Encoders`.  For instance: `userIdTextEncoder = tshowEncoder . unwrappedEncoder`.  This would also have probably made it clearer that `greaterThanZeroEncoder` doesn't make sense.

Reply:
But `Prism`s are partial in one direction so there's never a guarantee of being able to roundtrip?

Exposing the constructor seems to have been my mistake, because you can create an invalid `UserId` and therefore an invalid route, I think. One function to create a valid `UserId` separate to any of the routing encoders might be enough.

I didn't want to use the `unwrappedEncoder` for my initial example as using that single line of template haskell to derive the `Wrapped` code involves turning on about six language extensions. Also using the `tshowEncoder` doesn't provide any control over what is an acceptable number for a `UserId`. 

I know it's a tutorial example but I'm trying to keep the complexity budget down and avoid the Haskell meme of "just trust me and turn on all these extensions". I'll come up with a cleaner and not-busted example. :)

Reply:
The roundtripping law is that each value you can encode must decode successfully, and to the same value that was originally encoded.  That does *not* imply (and it is not the case) that decoding must be total.  Instead, this creates a requirement that anything which fails to decode must also not be a possible output of `encode`.  It is also fine for multiple values may decode to the same value; exactly one of them will be the *canonical* encoding, and so we can always canonicalize encoded values (if they are valid) by doing `fmap encode . decode`.

Keeping complexity down is a great goal, but we need to be mindful of the other goals as well, one of which is correctness.  The reason this library exists is that it is hard to write correct route handling code, and this library is able to make it somewhat easier.  The value of `unwrappedEncoder` is that it is known to be actually correct.

`tshowEncoder` not providing control over what is an acceptable number for a UserId is exactly the point - it only does the read/show part, and doesn't need to understand anything else.  It's safe as long as read/show roundtrips.

It's important here to ensure we convey to the reader what the actual purpose of this library is.  If it were a good idea for them to build their route encoder and their route decoder separately, and then turn them into an Encoder, this library would be unnecessary.  The purpose of using the Encoder abstraction is that if you have two Encoders that you know are valid, you can put them together in a way that is *also* guaranteed to be valid.

Reply:
One decent heuristic to use here: `prism` should probably not appear in this tutorial, except possibly towards the end in a section about "how to unsafely build an encoder".